flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java

```diff
@@ -187,15 +187,21 @@ public void run() {
 				}
 			}
 
+			long lastTimeNanos = 0;
 			while (isRunning()) {
 				if (nextShardItr == null) {
 					fetcherRef.updateState(subscribedShardStateIndex, SentinelSequenceNumber.SENTINEL_SHARD_ENDING_SEQUENCE_NUM.get());
 
 					// we can close this consumer thread once we've reached the end of the subscribed shard
 					break;
 				} else {
-					if (fetchIntervalMillis != 0) {
-						Thread.sleep(fetchIntervalMillis);
+				if (fetchIntervalMillis != 0) {
+						long elapsedTimeNanos = System.nanoTime() - lastTimeNanos;
+						long sleepTimeMillis = fetchIntervalMillis - (elapsedTimeNanos / 1_000_000);
+						if(sleepTimeMillis > 0) {
+							Thread.sleep(sleepTimeMillis);
+						}
+						lastTimeNanos = System.nanoTime();
```

Comment:
repeated call to nanoTime()

Reply:
@jgrier and I stepped through this and using a single call to nanoTime() will not account for timing the duration we sleep. leaving as is.